### PR TITLE
PLT-724 - Add action for syncing an image between ECRs

### PR DESCRIPTION
**Pull request by jake-naughton**
- This will be used as part of Fargate deployments to Production.
- Tested and working: https://github.com/cartoncloud/webapp-php/actions/runs/9459709376/job/26057337101

### CODE REVIEW

**Review by aarontwf — APPROVED** [review]
Guessing this is just for PHP. Multi-arch images (i.e ones supporting both AMD and ARM) may be more complex, so may need an update when adding that support to PHP.

**Comment by jake-naughton** [comment]
> Guessing this is just for PHP. Multi-arch images (i.e ones supporting both AMD and ARM) may be more complex, so may need an update when adding that support to PHP.

After PHP gets shifted to Fargate we will be moving Java and it will be following similar principles i.e. separating prod and SDLC ECRs. So it will end up using the same action eventually.

Do we need that for SDLC + prod environments? I thought that would only be for local environments which is separate again (no need to copy those through the whole pipeline as prod or SDLC won't use them.) I assume we only need Multi-arch images for local development?

**Comment by aarontwf** [comment]
> > Guessing this is just for PHP. Multi-arch images (i.e ones supporting both AMD and ARM) may be more complex, so may need an update when adding that support to PHP.
> 
> After PHP gets shifted to Fargate we will be moving Java and it will be following similar principles i.e. separating prod and SDLC ECRs. So it will end up using the same action eventually.
> 
> Do we need that for SDLC + prod environments? I thought that would only be for local environments which is separate again (no need to copy those through the whole pipeline as prod or SDLC won't use them.) I assume we only need Multi-arch images for local development?

It currently is using multi-arch everywhere so that local uses identical images.

Done here if interested
- https://github.com/cartoncloud/actions/blob/v3/build-java/action.yml#L130

**Review by tomaszCarton — APPROVED** [review]

**Comment by jake-naughton** [comment]
> > > Guessing this is just for PHP. Multi-arch images (i.e ones supporting both AMD and ARM) may be more complex, so may need an update when adding that support to PHP.
> > 
> > 
> > After PHP gets shifted to Fargate we will be moving Java and it will be following similar principles i.e. separating prod and SDLC ECRs. So it will end up using the same action eventually.
> > Do we need that for SDLC + prod environments? I thought that would only be for local environments which is separate again (no need to copy those through the whole pipeline as prod or SDLC won't use them.) I assume we only need Multi-arch images for local development?
> 
> It currently is using multi-arch everywhere so that local uses identical images.
> 
> Done here if interested
> 
> * https://github.com/cartoncloud/actions/blob/v3/build-java/action.yml#L130

I'm sure we can sort it out anyway when the time comes, are you saying the build process for Multi-arch produces different images to what you get without it? Otherwise, I don't really see the point. They should be identical. When building for an image for local you could include another platform 🤷‍♂️ Or we can leave it in I guess and just pay the extra storage + space that is required even though it wouldn't be used by production.

Maybe we could build this way in SDLC and then just pass the required platform through to Prod.

`local uses identical images` - Not when you're using different architectures, you are pulling different images anyway?

File: sync-image-between-ecrs/action.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      env:
+        FROM_REGISTRY: ${{ inputs.fromEcrRegistry }}
+        FROM_REPOSITORY: ${{ inputs.fromEcrRepo }}
+        FROM_IMAGE_TAG: ${{ inputs.fromImageTag }}
+        TO_REGISTRY: ${{ inputs.toEcrRegistry }}
+        TO_REPOSITORY: ${{ inputs.toEcrRepo }}
+        TO_IMAGE_TAG: ${{ inputs.toImageTag }}
+        VERSION_TAG: ${{ inputs.versionTag }}
+      run: |
+        docker tag $FROM_REGISTRY/$FROM_REPOSITORY:$FROM_IMAGE_TAG $TO_REGISTRY/$TO_REPOSITORY:$FROM_IMAGE_TAG
```
Comment (aarontwf) [inline sync-image-between-ecrs/action.yml:73]: This one may be a bit odd if you're not familiar with this action. What cases is this used for?

**Comment by aarontwf** [comment]
> > > > Guessing this is just for PHP. Multi-arch images (i.e ones supporting both AMD and ARM) may be more complex, so may need an update when adding that support to PHP.
> > > 
> > > 
> > > After PHP gets shifted to Fargate we will be moving Java and it will be following similar principles i.e. separating prod and SDLC ECRs. So it will end up using the same action eventually.
> > > Do we need that for SDLC + prod environments? I thought that would only be for local environments which is separate again (no need to copy those through the whole pipeline as prod or SDLC won't use them.) I assume we only need Multi-arch images for local development?
> > 
> > 
> > It currently is using multi-arch everywhere so that local uses identical images.
> > Done here if interested
> > 
> > * https://github.com/cartoncloud/actions/blob/v3/build-java/action.yml#L130
> 
> I'm sure we can sort it out anyway when the time comes, are you saying the build process for Multi-arch produces different images to what you get without it? Otherwise, I don't really see the point. They should be identical. When building for an image for local you could include another platform 🤷‍♂️ Or we can leave it in I guess and just pay the extra storage + space that is required even though it wouldn't be used by production.
> 
> Maybe we could build this way in SDLC and then just pass the required platform through to Prod.

Can't remember fully how it works but I think in ECR it shows 3 images, i.e. `1.2.3` is an image that simply points to either of the other two based on your arch.

> `local uses identical images` - Not when you're using different architectures, you are pulling different images anyway?

Was more meaning exact same build process and pointing to same tag

**Comment by jake-naughton** [comment]
> Was more meaning exact same build process and pointing to same tag

To be honest, I don't think much work would be required to get this working anyway when the time comes if we feel this is the right approach 👍

File: sync-image-between-ecrs/action.yml (hunk 1 above)
Comment (jake-naughton) [reply sync-image-between-ecrs/action.yml:73]: Have added a comment explaining why this is necessary.

**Pull request merged**